File: backend/shared/domain/primitives/domain_event.py
from __future__ import annotations

from dataclasses import dataclass
from datetime import datetime, timezone
from uuid import UUID, uuid4

from .unique_id import UniqueId
from .tenant_id import TenantId


@dataclass(frozen=True, slots=True)
class EventMetadata:
    correlation_id: UUID
    tenant_id: TenantId
    occurred_at: datetime
    causation_id: UUID | None = None
    user_id: UUID | None = None

    @classmethod
    def create(
        cls,
        tenant_id: TenantId,
        correlation_id: UUID | None = None,
        user_id: UUID | None = None,
        causation_id: UUID | None = None,
    ) -> EventMetadata:
        return cls(
            correlation_id=correlation_id or uuid4(),
            tenant_id=tenant_id,
            occurred_at=datetime.now(timezone.utc),
            causation_id=causation_id,
            user_id=user_id,
        )


class DomainEvent:
    """Internal domain event — never leaves the bounded context boundary directly."""

    __slots__ = ("event_id", "aggregate_id", "metadata", "event_version")

    def __init__(
        self,
        aggregate_id: UniqueId,
        metadata: EventMetadata,
        event_version: int = 1,
        event_id: UniqueId | None = None,
    ) -> None:
        self.event_id = event_id or UniqueId.generate()
        self.aggregate_id = aggregate_id
        self.metadata = metadata
        self.event_version = event_version

    @property
    def event_name(self) -> str:
        raise NotImplementedError

    def to_integration_payload(self) -> dict:
        raise NotImplementedError

File: packages/shared-kernel/src/domain/domain-event.ts
import { UniqueEntityId } from "./unique-entity-id";
import { TenantId } from "./tenant-id";

export interface DomainEventMetadata {
  readonly correlationId: string;
  readonly causationId?: string | undefined;
  readonly tenantId: TenantId;
  readonly userId?: string | undefined;
  readonly occurredAt: Date;
}

export abstract class DomainEvent {
  readonly eventId: UniqueEntityId;
  readonly aggregateId: UniqueEntityId;
  readonly metadata: DomainEventMetadata;
  abstract readonly eventName: string;
  abstract readonly eventVersion: number;

  protected constructor(
    aggregateId: UniqueEntityId,
    metadata: DomainEventMetadata,
    eventId?: UniqueEntityId,
  ) {
    this.eventId = eventId ?? UniqueEntityId.create();
    this.aggregateId = aggregateId;
    this.metadata = metadata;
  }

  toPayload(): Record<string, unknown> {
    return {
      eventId: this.eventId.toString(),
      eventName: this.eventName,
      eventVersion: this.eventVersion,
      aggregateId: this.aggregateId.toString(),
      metadata: {
        correlationId: this.metadata.correlationId,
        causationId: this.metadata.causationId,
        tenantId: this.metadata.tenantId.toString(),
        userId: this.metadata.userId,
        occurredAt: this.metadata.occurredAt.toISOString(),
      },
    };
  }
}
